phoenix-core/src/main/java/org/apache/phoenix/expression/function/PhoenixRowTimestampFunction.java

```diff
@@ -124,7 +127,7 @@ public boolean isStateless() {
 
     @Override
     public Determinism getDeterminism() {
-        return Determinism.PER_ROW;
+        return Determinism.ALWAYS;
```

Comment:
Why is this needed? And is it correct? (The row timestamp is per row.)

Reply:
There is a check in MetaDataClient#createIndex(), which does not allow non-deterministic expressions in index and considers that anything other than Determinism.ALWAYS is non-deterministic. This does not seem right. SQL Server agrees with me "Deterministic functions always return the same result any time they are called with a specific set of input values and given the same state of the database. Nondeterministic functions may return different results each time they are called with a specific set of input values even if the database state that they access remains the same. For example, the function AVG always returns the same result given the qualifications stated above, but the GETDATE function, which returns the current datetime value, always returns a different result." Maybe I need to change the create index code to allow PER_ROW determinism too. What do you say?

Reply:
Hmm... Interesting. If you do a transaction (or query) over 1000 rows you would not expect that PHOENIX_ROW_TIMESTAMP() returning the same one value, so ALWAYS seems wrong.

Doing a git blame it looks like I, myself, added PER_INVOCATION with PHOENIX-1516, prior to that there was only ROW, STATEMENT, and ALWAYS.

I agree with you that indexes should be fine with PER_ROW and ALWAYS determinism.
PER_INVOCATION should be disallowed. Not 100% sure about PER_STATEMENT, seems that would cause problems. It is possible that some function are incorrectly labeled, though, we do not seems to follow that logic correctly.